### ISSUE TITLE
Improve PyTorch classifier predict speed by not calculating gradients

### ISSUE THREAD

**Issue by davidslater** [IBM/adversarial-robustness-toolbox#448]
When running classifier.predict on a pytorch wrapper, gradients are calculated and then thrown out. By not calculating gradients, it should give a 2-3x speedup on predict.

The solution is pretty straightforward - wrap the model predict with
```
with torch.no_grad():
    ...
```
or just wrap this single line:
https://github.com/IBM/adversarial-robustness-toolbox/blob/3cfa6de196cb32a3efafab2ff6bbf44247c9ddbd/art/classifiers/pytorch.py#L147

This was a PR in ARMORY - https://github.com/twosixlabs/armory/pull/608 - but I think it makes more sense in ART. Let me know your thoughts.

**Comment by beat-buesser** [comment]
@davidslater Thank you for raising this issue! I think this makes sense and we should include it in ART 1.3. I would chose wrapping the single line 147.

Would you like to make a PR with a solution as mentioned in your twosixlabs/armory#608 or should we implemented it?

**Comment by davidslater** [comment]
I am fine making a PR for it. Could you point to any particular ways you would like me to benchmark the speed differences?

**Comment by beat-buesser** [comment]
I think the advantage of `torch.no_grad()` is clearly documented. If you have time for a benchmark, I would probably use a model like `torchvision.models.resnet50` in `PyTorchClassifier` and test it on a batch of images.
I think this is a great catch, especially for black-box attacks.

**Comment by davidslater** [comment]
#452 resolves the issue.
